### PR TITLE
Add a type class for measures

### PR DESCRIPTION
**Pull request by nfrisby**
See the individual commits and their messages.

### CODE REVIEW

File: measures/src/Data/Measure.hs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module Data.Measure (
```
Comment (nfrisby) [inline measures/src/Data/Measure.hs:5]: Add comment here

File: base-deriving-via/src/Data/DerivingVia/GHC/Generics/Semigroup.hs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  gsappend :: rep x -> rep x -> rep x
+
+instance Monoid c => GSemigroup (K1 i c) where
+  gsappend (K1 l) (K1 r) = K1 (l <> r)
+
+instance GSemigroup f => GSemigroup (M1 i c f) where
+  gsappend (M1 l) (M1 r) = M1 (gsappend l r)
+
+instance GSemigroup V1 where
+  gsappend = error "GSemigroup V1"
```
Comment (coot) [inline base-deriving-via/src/Data/DerivingVia/GHC/Generics/Semigroup.hs:28]: Should we use a `TypeError` here as well?

Comment (coot) [reply base-deriving-via/src/Data/DerivingVia/GHC/Generics/Semigroup.hs:28]: Actually it shouldn't be an error, `Void` has a valid `Semigroup` instance.

Comment (nfrisby) [reply base-deriving-via/src/Data/DerivingVia/GHC/Generics/Semigroup.hs]: Thanks @coot; that was a copy-paste error. Nice catch.

**Review by dcoutts — APPROVED** [review]

**Review by nc6 — APPROVED** [review]
LGTM

**Comment by nfrisby** [comment]
My latest updates were because I didn't realize how many times I had replicated the mindless `V1` mistake that @coot caught. I think I got them all now.

**Comment by nc6** [comment]
bors merge

**Pull request merged**